File: python-sdk/visa_direct_sdk/utils/events.py
from typing import Any, Dict


class CompensationEventEmitter:

	async def emit(self, event: Dict[str, Any]) -> None:  # pragma: no cover
		raise NotImplementedError


class LogEmitter(CompensationEventEmitter):

	async def emit(self, event: Dict[str, Any]) -> None:  # pragma: no cover
		if not validate_compensation_event(event):
			raise ValueError("Invalid compensation event payload")
		print('[compensation]', {k: event.get(k) for k in ('event', 'sagaId', 'reason', 'timestamp')})


def validate_compensation_event(event: Dict[str, Any]) -> bool:
	have_required = all(event.get(field) for field in ("event", "sagaId", "reason", "timestamp"))
	return bool(have_required and event.get("event") == "payout_failed_requires_compensation")

File: typescript-sdk/src/utils/events.ts
export type CompensationEvent = {
	event: 'payout_failed_requires_compensation';
	sagaId: string;
	funding: any;
	reason: string;
	metadata?: any;
	timestamp: string;
};

export interface CompensationEventEmitter {
	emit(evt: CompensationEvent): Promise<void>;
}

export class LogEmitter implements CompensationEventEmitter {
	async emit(evt: CompensationEvent): Promise<void> {
		// Validate schema before emit
		if (!evt.event || !evt.sagaId || !evt.reason || !evt.timestamp) {
			throw new Error('Invalid compensation event schema');
		}
		// Avoid PII; log minimal redacted info
		console.warn('[compensation]', { event: evt.event, sagaId: evt.sagaId, reason: evt.reason, timestamp: evt.timestamp });
	}
}

// Schema validator for compensation events
export function validateCompensationEvent(evt: any): evt is CompensationEvent {
	return evt && 
		typeof evt.event === 'string' &&
		typeof evt.sagaId === 'string' &&
		typeof evt.reason === 'string' &&
		typeof evt.timestamp === 'string' &&
		evt.event === 'payout_failed_requires_compensation';
}


